### PR TITLE
Stale Case Data Mail Report

### PR DESCRIPTION
**Pull request by zandre-eng**
## Product Description

No user-facing changes. This new report is only sent out via mail and is not included in HQ's UI.

## Technical Summary

Link to ticket [here](https://dimagi.atlassian.net/browse/SC-4039).

A new periodic Celery task is created with the responsibility of sending out a email report to the `SOLUTIONS_AES_EMAIL` email address on a monthly basis. This is a simple report that aggregates case counts for domains that have open cases that have not been modified in over a year.

The query logic itself has date chunking implemented, which essentially means the queries will be made in date slices of 180 days at a time. The slices will go back 20 years in these slices. This specific date has been picked as prod testing has shown that this is about where the oldest cases reside.

Additionally, the query logic also implements backoff logic. Essentially, if the query times out then the date slice will be made thinner by 30 days. If this happens more than 2 times then the table will stop aggregating data entirely and will simply return the data it has aggregated so far. This logic has been implemented specifically with production's scale in mind, as the query can sometimes time out on the 1-2 most recent time slices due to the amount of data there is to aggregate.

The goal of this report is to identify which domains could be potential candidates for auto-update rules to be implemented.

## Feature Flag

None

### Safety story

- Local testing
- Unit tests

The Celery task has also been tested on the staging environment to ensure that the email and attachment are correctly sent.

### Automated test coverage

Unit tests exist for the new `StaleCasesTable` class and test to make sure that it retrieves the correct data and formats it appropriately.

### QA Plan

No QA planned.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by orangejenny** [comment]
Code looks fine, though I only skimmed it. Do you know how many domains currently meet this criteria on prod? I wonder if it would be helpful to filter them, possibly by subscription, to domains that solutions is most likely to take action on.

File: corehq/apps/hqadmin/reports.py [hunk 1]
```diff
@@ -495,3 +498,64 @@ def headers(self):
...
+
+    @property
+    def headers(self):
+        return DataTablesHeader(
+            DataTablesColumn(gettext_lazy("Domain")),
+            DataTablesColumn(gettext_lazy("Case count"))
+        )
+
+    @property
+    def rows(self):
```
Comment (kaapstorm) [inline corehq/apps/hqadmin/reports.py:515]: Using `table.rows` twice will run the query twice. How about something like this?
```python
    def __init__():
        self._rows = None

    @property
    def rows(self):
        if self.rows is None:
            self._rows = []
            # etc.
        return self._rows
```

File: corehq/apps/hqadmin/reports.py [hunk 2]
```diff
@@ -495,3 +498,64 @@ def headers(self):
...
+            )
+        ).run().aggregations.domain.buckets_dict
+
+    def _get_stale_date(self):
+        current_date = datetime.now()
+        stale_threshold_date = current_date - timedelta(days=self.STALE_DATE_THRESHOLD_DAYS)
+        return stale_threshold_date
+
+    @staticmethod
+    def format_as_table(row_data, data_tables_header):
```
Comment (kaapstorm) [inline corehq/apps/hqadmin/reports.py:539]: This is a useful function, but it is more generic than this class, and it expects its headers to be a list of DataTablesHeader instances. For those reasons, I would pull it out of here, move it somewhere under `corehq/apps/reports/datatables/`, and maybe rename it. If you feel that having a method is useful, you could change this to something like:
```python
    def format_as_table(self):
        return datatable_as_text(self.rows, self.headers)
```
(Obviously assumes that `self.rows` doesn't run the query again.)

(Not applicable here because it's probably not worth an additional requirement, but useful to know about: [Tablib](https://tablib.readthedocs.io/en/stable/) is cool, and [it can do this](https://tablib.readthedocs.io/en/stable/formats.html#rst). (Also, I'm biased, because I wrote that bit.))

**Review by kaapstorm — COMMENTED** [review]
This is nice. Just a couple of suggestions.

File: corehq/apps/hqadmin/reports.py (hunk 1 above)
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py:515]: I like this idea! Currently the "caching" is done in the task itself by calling `table_rows = StaleCasesTable().rows` and then using the `table_rows` var locally, but the above suggestion will remove the need for storing the rows locally in the task.

File: corehq/apps/hqadmin/reports.py (hunk 2 above)
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py:539]: That is a good point. We no longer need this function as I'll be refactoring the task to save the data to a csv file instead. I'm therefore going to completely remove this method. 
I thought it might be useful to keep it around and put it in `corehq/apps/reports/datatables` as you suggested for future use but, in saying that, YAGNI.

**Review by ajeety4 — COMMENTED** [review]
Looks good to me.

**Comment by zandre-eng** [comment]
> Do you know how many domains currently meet this criteria on prod?

@orangejenny There are about 14k domains that meet the criteria on prod.

> I wonder if it would be helpful to filter them, possibly by subscription, to domains that solutions is most likely to take action on.

This is a great suggestion. Implementing filters for active subscriptions only and excluding the community plan domains drops the total count from 14k to about 4k.

File: corehq/apps/hqwebapp/tasks.py [hunk 3]
```diff
@@ -305,3 +310,55 @@ def send_domain_ucr_data_info_to_admins():
...
+        writer = csv.writer(csv_file)
+        writer.writerow(table.headers)
+        writer.writerows(table.rows)
+    else:
+        message = (
+            'No domains were found containing case data older than '
+            f'{table.STALE_DATE_THRESHOLD_DAYS} days.'
+        )
+        if has_error:
+            message += (
```
Comment (ajeety4) [inline corehq/apps/hqwebapp/tasks.py:355]: nit: In this scenario, I think it is probably better to emphasize on the error message first and then speak about no domains found in the partial compilation (or just keep the error message). I say this because I am not sure how relevant is the above message to the end user in case of error.

File: corehq/apps/hqadmin/reports.py [hunk 4]
```diff
@@ -495,3 +502,92 @@ def headers(self):
...
+        return self._rows
+
+    def _aggregate_case_count_data(self):
+        end_date = datetime.now() - timedelta(days=self.STALE_DATE_THRESHOLD_DAYS)
+        agg_res = {}
+        curr_backoff_count = 0
+        curr_agg_date_range = self.AGG_DATE_RANGE
+        domains = self._get_domains()
+        while (True):
+            start_date = end_date - timedelta(days=self.AGG_DATE_RANGE)
```
Comment (ajeety4) [inline corehq/apps/hqadmin/reports.py:542]: Should this be `curr_agg_date_range` instead of `self.AGG_DATE_RANGE` ?

File: corehq/apps/hqadmin/reports.py [hunk 5]
```diff
@@ -495,3 +502,92 @@ def headers(self):
...
+            try:
+                query_res = self._stale_case_count_in_date_range(domains, start_date, end_date)
+            except ESError as e:
+                curr_backoff_count += 1
+                if curr_backoff_count <= self.MAX_BACKOFF_COUNT:
+                    curr_agg_date_range -= self.BACKOFF_AMOUNT_DAYS
+                    curr_backoff_count += 1
```
Comment (ajeety4) [inline corehq/apps/hqadmin/reports.py:549]: Looks like we are incrementing here it again ?

File: corehq/apps/hqadmin/reports.py [hunk 6]
```diff
@@ -495,3 +502,92 @@ def headers(self):
...
+                        None,
+                        'ES query timed out while compiling stale case report email.',
+                        details={
+                            'error': str(e),
+                            'start_date': start_date.strftime("%Y-%m-%d"),
+                            'end_date': end_date.strftime("%Y-%m-%d")
+                        }
+                    )
+                    raise ESError()
+            curr_backoff_count = 0
```
Comment (ajeety4) [inline corehq/apps/hqadmin/reports.py:561]: I might be wrong here however thie block from here onwards seems like it should be in the else block of try.

**Comment by ajeety4** [comment]
> To make the ES query performant enough and consistently successful I have refactored how the report does its query a bit. Now the ES query will aggregate data in date slices of 180 days at a time. I have also implemented backoff logic so that the date slices become thinner if the query times out. When testing on production I have noted that this may happen and looks to be dependent on whether the index has been cached in some way. The worst case scenario I have witnessed for production is that it might time out the first 1-2 queries before successfully completing all of them.
> 
> Please do let me know if you have any thoughts/concerns with the above strategy, and am happy to discuss alternatives as well.

A great idea to split this into days.  The other alternative to split is probably domains however days seems like a better candidate to achieve an even distribution for the split.  Good thinking on having the backoff strategy to make it more reliable. 
I don't foresee any concerns and this does seem like a ramp up to the previous approach.

File: corehq/apps/hqwebapp/tasks.py (hunk 3 above)
Comment (zandre-eng) [reply corehq/apps/hqwebapp/tasks.py:355]: That is a good point. This could potentially be confusing to have a message that says no domains found followed by one that says there was an error. I'll refactor this so that the default no domains message is only shown if there wasn't an error.

File: corehq/apps/hqadmin/reports.py (hunk 5 above)
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py:549]: Ah, yeah we should only be incrementing outside the conditional block. Will remove the other one.

File: corehq/apps/hqadmin/reports.py (hunk 6 above)
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py:561]: That is a good point. In my experimental script code I had a continue in the except block which seems I didn't carry over to here. But I like the idea of using an else block rather. Will implement that here.

**Comment by mkangia** [comment]
Hey @zandre-eng 

Just checking if we considered the opposite appraoch?
fetch domains that have a case update in last one year and then filter those from a list of relevant domains like the ones that have an active subscription?

File: corehq/apps/hqwebapp/tasks.py (hunk 3 above)
Comment (zandre-eng) [reply corehq/apps/hqwebapp/tasks.py]: Addressed in [ee92b43](https://github.com/dimagi/commcare-hq/pull/35446/commits/ee92b43c1856f7312e6d64c1fdf6d205104a21f0)

File: corehq/apps/hqadmin/reports.py (hunk 4 above)
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py]: Addressed in [b228b40](https://github.com/dimagi/commcare-hq/pull/35446/commits/b228b408ee63cb87d5786417b375e517c54afcd6)

File: corehq/apps/hqadmin/reports.py (hunk 5 above)
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py:548]: Addressed in [b228b40](https://github.com/dimagi/commcare-hq/pull/35446/commits/b228b408ee63cb87d5786417b375e517c54afcd6)

File: corehq/apps/hqadmin/reports.py (hunk 6 above)
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py]: Addressed in [b228b40](https://github.com/dimagi/commcare-hq/pull/35446/commits/b228b408ee63cb87d5786417b375e517c54afcd6)

**Comment by zandre-eng** [comment]
> Just checking if we considered the opposite approach?
fetch domains that have a case update in last one year and then filter those from a list of relevant domains like the ones that have an active subscription?

@mkangia Testing this out the query times out when trying to fetch all domans that have had a case update in the last year. To be able to fetch this we would need to implement date chunking here as well and I believe the complexity/cost of implementing these further additional queries outweighs the benefits we get from filtering out these extra domains.

File: corehq/apps/hqadmin/reports.py (hunk 5 above)
Comment (ajeety4) [reply corehq/apps/hqadmin/reports.py:548]: looks like the wrong one was removed here

Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py]: :man_facepalming: Apologies silly mistake on my end. I did a rebase and put it with the previous commit in [7ec4e5b](https://github.com/dimagi/commcare-hq/pull/35446/commits/7ec4e5b3a68e7c647c8dd70613597a1667d59b8b)

**Comment by mkangia** [comment]
Hey @zandre-eng 

> Testing this out the query times out when trying to fetch all domans that have had a case update in the last year.

oh wow! I thought that that would perform better.
I see that you have tried to make things work in the best possible way with case ES. It does appear it will continue to face challenges as we get more and more cases in the system.

One suggestion I could think of is to instead use forms.
It seems FormES already supports filtering by submission date as seen [here](https://github.com/dimagi/commcare-hq/blob/3cc9c98baaa9a1a0a36eaa0d9f0d0c1062de06cb/corehq/apps/es/forms.py#L185) so it should work better.
May be its worth give it a go.

I believe you could also make this better for either approach by iterating with [active subscriptions](https://github.com/dimagi/commcare-hq/blob/b0ab4b1b281fb84ddcf1b3eb444e2d3a778ba46f/corehq/apps/accounting/models.py#L1192), and then [find domains for those](https://github.com/dimagi/commcare-hq/blob/b0ab4b1b281fb84ddcf1b3eb444e2d3a778ba46f/corehq/apps/accounting/models.py#L1901), and then look for their last case update or form submission flagging them as inactive if needed.
Basically, instead of fetching information from all cases, instead fetch it only for relevant domains/subscriptions.

Feel free to go with the approach currently in the PR if you think that will work well or keep them as a follow up. It does seem the requirement here is to find inactive domains and that can be done in different ways.

**Comment by zandre-eng** [comment]
> oh wow! I thought that that would perform better.

@mkangia Docs in the `case_search` index are generally larger and more complex than that of the `case` index, so my suspicion is that this seems to be a big enough difference to affect performance quite a bit.

> One suggestion I could think of is to instead use forms.
It seems FormES already supports filtering by submission date as seen [here](https://github.com/dimagi/commcare-hq/blob/3cc9c98baaa9a1a0a36eaa0d9f0d0c1062de06cb/corehq/apps/es/forms.py#L185) so it should work better.
May be its worth give it a go.

We could query using `FormES` with something like:
```
stale_date = datetime.now() - timedelta(days=365)
(
    FormES()
    .submitted(lt=stale_date)
    .aggregation(
        TermsAggregation('domain', 'domain.exact')
    )
    .size(0)
).run().aggregations.domain.counts_by_bucket().keys()
```
The only problem with the above however, is that we can't easily filter out forms that affect currently closed cases. This means querying for forms older than a year will return _a lot more_ domains than we're interested in. We would need to further filter down this list by some other means, such as using the `Subscription` model. At that point, I feel the form query isn't adding much benefit.

> I believe you could also make this better for either approach by iterating with [active subscriptions](https://github.com/dimagi/commcare-hq/blob/b0ab4b1b281fb84ddcf1b3eb444e2d3a778ba46f/corehq/apps/accounting/models.py#L1192), and then [find domains for those](https://github.com/dimagi/commcare-hq/blob/b0ab4b1b281fb84ddcf1b3eb444e2d3a778ba46f/corehq/apps/accounting/models.py#L1901), and then look for their last case update or form submission flagging them as inactive if needed.
Basically, instead of fetching information from all cases, instead fetch it only for relevant domains/subscriptions.

Would this not significantly increase the amount of queries we would need to do? The `Subscription.get_active_domains_for_account()` function does a query to fetch all the active subscription domains for an account. For each of the accounts containing active subscription domains we would need to run an aggregation query to identify which domains contain stale cases.

If we had 100 active accounts this would already result in 200 queries, which is more than the ~38 we're doing now. Please do let me know though if I misunderstood regarding the above.

**Review by kaapstorm — APPROVED** [review]
All good! Just a suggestion.

File: corehq/apps/hqadmin/reports.py [hunk 7]
```diff
@@ -495,3 +502,92 @@ def headers(self):
...
+        if self._rows is None:
+            self._rows = []
+            case_count_by_domain = self._aggregate_case_count_data()
+            for domain, case_count in case_count_by_domain.items():
+                self._rows.append([domain, case_count])
+        return self._rows
+
+    def _aggregate_case_count_data(self):
+        end_date = datetime.now() - timedelta(days=self.STALE_DATE_THRESHOLD_DAYS)
+        agg_res = {}
```
[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py:537]: This is a nice improvement! I completely forgot about the handy `defaultdict` type. I'll add this in.

[the remark this answers is not in the document]
Comment (zandre-eng) [reply corehq/apps/hqadmin/reports.py]: Addressed in [25d9dc5](https://github.com/dimagi/commcare-hq/pull/35446/commits/25d9dc5cffb9417c6873aaf7a86cf7afb715e134)

**Review by kaapstorm — APPROVED** [review]

File: corehq/apps/hqadmin/reports.py (hunk 5 above)
Comment (ajeety4) [reply corehq/apps/hqadmin/reports.py]: No worries

**Comment by mkangia** [comment]
Hey @zandre-eng 

> The only problem with the above however, is that we can't easily filter out forms that affect currently closed cases

Do you mean system forms?
If yes, those probably can be filtered out since they have the same XMLNS.

> Would this not significantly increase the amount of queries we would need to do? The Subscription.get_active_domains_for_account() function does a query to fetch all the active subscription domains for an account. For each of the accounts containing active subscription domains we would need to run an aggregation query to identify which domains contain stale cases.

That is correct. More queries that could run fast and effectively. I am assuming that running through ES for a full search at once is going to be a pain over time.

I'll be honest that I have not considered all variables of this so I am just suggesting things from a high level understanding. Feel free to keep the original implementation if that is a better one.

**Comment by mkangia** [comment]
Sounds good to me @zandre-eng :+1: 
Thanks for discussing the approach
